context.go

```diff
@@ -0,0 +1,106 @@
+// https://www.xbrl.org/Specification/XBRL-2.1/REC-2003-12-31/XBRL-2.1-REC-2003-12-31+corrected-errata-2013-02-20.html#_4.7.3
+type Entity struct {
+	Identifier Identifier `xml:"identifier"`
+	Segments   Segments   `xml:"segment"`
+}
+
+// Identifier specifies a scheme for identifying business entities and an identifier that follows the scheme.
+// https://www.xbrl.org/Specification/XBRL-2.1/REC-2003-12-31/XBRL-2.1-REC-2003-12-31+corrected-errata-2013-02-20.html#_4.7.3.1
+// For Example:
+// <identifier scheme="http://www.sec.gov/CIK">0000320193</identifier>
+//
+// The above `identifier` element specifies that the scheme for identifying the entity is through SEC CIK numbers,
+// and that the identifier itself is 0000320193 (the CIK for Apple Inc.).
+type Identifier struct {
+	Scheme string `xml:"scheme,attr"`
+	Value  string `xml:",chardata"`
+}
+
+// Segments is a type alias for a slice of Segment structs.
+// It implements xml.Unmarshaller and puts and unmarshals any sub-elements as a Segment and puts it into the slice.
+type Segments []Segment
+
+// Segment is an optional container for additional information used to identify a business segment more completely
+// for cases where the Identifier is insufficient.
+// There are no Segments defined in the base XBRL spec, they must be defined in other XML schemas.
+// https://www.xbrl.org/Specification/XBRL-2.1/REC-2003-12-31/XBRL-2.1-REC-2003-12-31+corrected-errata-2013-02-20.html#_4.7.3.2
+type Segment struct {
+	XMLName    xml.Name
+	Attributes []xml.Attr `xml:",any,attr"`
+	Value      string     `xml:",chardata"`
+}
+
+// UnmarshalXML implements xml.Unmarshaller for Segments.
+// It unmarshals any sub-elements as Segments and puts them into this slice.
+func (s *Segments) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	var segmentsAnon struct {
+		Segments []Segment `xml:",any"`
+	}
+
+	if err := d.DecodeElement(&segmentsAnon, &start); err != nil {
+		return err
+	}
+
+	*s = segmentsAnon.Segments
+	return nil
+}
+
+type PeriodType string
+
+// All the supported PeriodType values. See Period.Type() for more information.
+const (
+	Duration PeriodType = "duration"
+	Instant  PeriodType = "instant"
+	Forever  PeriodType = "forever"
+	Invalid  PeriodType = "invalid"
+)
+
+// Period contains an instant or interval of time for a Context.
+// https://www.xbrl.org/Specification/XBRL-2.1/REC-2003-12-31/XBRL-2.1-REC-2003-12-31+corrected-errata-2013-02-20.html#_4.7.2
+type Period struct {
+	// StartDate is non-nil and guaranteed to be before EndDate if Period.Type() returns Duration.
+	StartDate *string `xml:"startDate"`
+	// EndDate is non-nil and guaranteed to be after StartDate if Period.Type() returns Duration.
+	EndDate *string `xml:"endDate"`
+
+	// Instant is non-nil if Period.Type() returns Instant
+	Instant *string `xml:"instant"`
+
+	// Forever is non-nil if Period.Type() returns Forever.
+	// Note ideally this would be a bool, but the XML Unmarshaller doesn't support
```

Comment:
Should we make a custom unmarshal for it?

Reply:
I don't think it's really necessary. `Type()` returning `Forever` is already a clean way to determine if the period represents 'forever', so you wouldn't really have to interact with this property directly as a user. 

I might come back to this later once I'm more experienced with custom unmarshallers though